File: modules/swinglink-ble/ios/BLEScanner.swift
import CoreBluetooth
import Foundation

/// Scans for nearby SwingLink BLE advertisers and reads their payloads via GATT.
///
/// Discovery flow for each new peripheral:
///   1. `didDiscover` — cache RSSI, connect for GATT read
///   2. `didConnect` → `discoverServices` → `discoverCharacteristics` → `readValue`
///   3. Emit `onDeviceFound` with parsed payload, disconnect
///
/// Subsequent sightings of the same peripheral skip GATT (use cached payload),
/// re-emitting `onDeviceFound` with updated RSSI.
///
/// A staleness timer fires every 3 seconds. Devices not seen for 10 seconds
/// emit `onDeviceLost` and are purged from the cache.
class BLEScanner: NSObject, CBCentralManagerDelegate, CBPeripheralDelegate {

  /// Called when a device is discovered or its RSSI/lastSeen is updated.
  var onDeviceFound: (([String: Any]) -> Void)?
  /// Called when a device goes stale and is removed.
  var onDeviceLost: ((String) -> Void)?

  private var centralManager: CBCentralManager?
  private var stalenessTimer: Timer?

  // Peripherals currently undergoing GATT read (held to prevent dealloc)
  private var pendingPeripherals: [UUID: CBPeripheral] = [:]

  // Cache of successfully read devices: peripheral UUID -> device info dict
  private var readDevices: [UUID: [String: Any]] = [:]

  // Timestamp tracking for staleness
  private var lastSeenTimestamps: [UUID: TimeInterval] = [:]

  // RSSI cache for peripherals (updated on every advertisement)
  private var rssiCache: [UUID: NSNumber] = [:]

  private let stalenessInterval: TimeInterval = 3.0
  private let stalenessThreshold: TimeInterval = 10.0

  /// Starts scanning for SwingLink BLE advertisers.
  func startScanning() {
    if centralManager == nil {
      centralManager = CBCentralManager(delegate: self, queue: nil)
    } else if centralManager?.state == .poweredOn {
      beginScanning()
    }
  }

  /// Stops scanning, cancels pending GATT connections, and clears all caches.
  func stopScanning() {
    stalenessTimer?.invalidate()
    stalenessTimer = nil

    centralManager?.stopScan()

    for (_, peripheral) in pendingPeripherals {
      centralManager?.cancelPeripheralConnection(peripheral)
    }

    pendingPeripherals.removeAll()
    readDevices.removeAll()
    lastSeenTimestamps.removeAll()
    rssiCache.removeAll()
  }

  // MARK: - CBCentralManagerDelegate

  func centralManagerDidUpdateState(_ central: CBCentralManager) {
    if central.state == .poweredOn {
      beginScanning()
    }
  }

  func centralManager(
    _ central: CBCentralManager,
    didDiscover peripheral: CBPeripheral,
    advertisementData: [String: Any],
    rssi RSSI: NSNumber
  ) {
    let peripheralId = peripheral.identifier
    let now = Date().timeIntervalSince1970

    rssiCache[peripheralId] = RSSI
    lastSeenTimestamps[peripheralId] = now

    // If we already have a cached read for this device, re-emit with updated RSSI
    if var cached = readDevices[peripheralId] {
      cached["rssi"] = RSSI.intValue
      cached["lastSeen"] = now * 1000 // JS uses milliseconds
      readDevices[peripheralId] = cached
      onDeviceFound?(cached)
      return
    }

    // If already connecting/reading, skip
    if pendingPeripherals[peripheralId] != nil {
      return
    }

    // New peripheral — connect for GATT characteristic read
    pendingPeripherals[peripheralId] = peripheral
    peripheral.delegate = self
    central.connect(peripheral, options: nil)
  }

  func centralManager(_ central: CBCentralManager, didConnect peripheral: CBPeripheral) {
    peripheral.discoverServices([BLEConstants.serviceUUID])
  }

  func centralManager(
    _ central: CBCentralManager,
    didFailToConnect peripheral: CBPeripheral,
    error: Error?
  ) {
    pendingPeripherals.removeValue(forKey: peripheral.identifier)
  }

  // MARK: - CBPeripheralDelegate

  func peripheral(_ peripheral: CBPeripheral, didDiscoverServices error: Error?) {
    guard error == nil,
          let service = peripheral.services?.first(where: { $0.uuid == BLEConstants.serviceUUID })
    else {
      cleanupPeripheral(peripheral)
      return
    }
    peripheral.discoverCharacteristics([BLEConstants.payloadCharacteristicUUID], for: service)
  }

  func peripheral(
    _ peripheral: CBPeripheral,
    didDiscoverCharacteristicsFor service: CBService,
    error: Error?
  ) {
    guard error == nil,
          let characteristic = service.characteristics?.first(where: {
            $0.uuid == BLEConstants.payloadCharacteristicUUID
          })
    else {
      cleanupPeripheral(peripheral)
      return
    }
    peripheral.readValue(for: characteristic)
  }

  func peripheral(
    _ peripheral: CBPeripheral,
    didUpdateValueFor characteristic: CBCharacteristic,
    error: Error?
  ) {
    defer { cleanupPeripheral(peripheral) }

    guard error == nil,
          let data = characteristic.value,
          let payload = BLEConstants.unpackPayload(data)
    else { return }

    let peripheralId = peripheral.identifier
    let rssi = rssiCache[peripheralId]?.intValue ?? -100
    let now = Date().timeIntervalSince1970

    let deviceInfo: [String: Any] = [
      "id": peripheralId.uuidString,
      "name": peripheral.name as Any,
      "platform": payload.platform,
      "roomCode": payload.roomCode,
      "rssi": rssi,
      "lastSeen": now * 1000,
    ]

    readDevices[peripheralId] = deviceInfo
    lastSeenTimestamps[peripheralId] = now
    onDeviceFound?(deviceInfo)
  }

  // MARK: - Private

  private func beginScanning() {
    centralManager?.scanForPeripherals(
      withServices: [BLEConstants.serviceUUID],
      options: [CBCentralManagerScanOptionAllowDuplicatesKey: true]
    )
    startStalenessTimer()
  }

  private func cleanupPeripheral(_ peripheral: CBPeripheral) {
    centralManager?.cancelPeripheralConnection(peripheral)
    pendingPeripherals.removeValue(forKey: peripheral.identifier)
  }

  private func startStalenessTimer() {
    stalenessTimer?.invalidate()
    stalenessTimer = Timer.scheduledTimer(withTimeInterval: stalenessInterval, repeats: true) {
      [weak self] _ in
      self?.pruneStaleDevices()
    }
  }

  private func pruneStaleDevices() {
    let now = Date().timeIntervalSince1970
    var staleIds: [UUID] = []

    for (id, lastSeen) in lastSeenTimestamps {
      if now - lastSeen > stalenessThreshold {
        staleIds.append(id)
      }
    }

    for id in staleIds {
      let idString = readDevices[id]?["id"] as? String ?? id.uuidString
      readDevices.removeValue(forKey: id)
      lastSeenTimestamps.removeValue(forKey: id)
      rssiCache.removeValue(forKey: id)
      if let pending = pendingPeripherals.removeValue(forKey: id) {
        centralManager?.cancelPeripheralConnection(pending)
      }
      onDeviceLost?(idString)
    }
  }
}

File: modules/divot-ble/android/src/main/java/com/divotgolf/ble/BLEScanner.kt
package com.divotgolf.ble

import android.bluetooth.BluetoothAdapter
import android.bluetooth.BluetoothGatt
import android.bluetooth.BluetoothGattCallback
import android.bluetooth.BluetoothGattCharacteristic
import android.bluetooth.BluetoothManager
import android.bluetooth.BluetoothProfile
import android.bluetooth.le.BluetoothLeScanner
import android.bluetooth.le.ScanCallback
import android.bluetooth.le.ScanFilter
import android.bluetooth.le.ScanResult
import android.bluetooth.le.ScanSettings
import android.content.Context
import android.os.Handler
import android.os.Looper
import android.util.Log

/**
 * Scans for nearby Divot BLE advertisers and reads their payloads.
 *
 * All advertisers (iOS and Android) serve the payload via a GATT readable
 * characteristic. On first discovery, the scanner connects via GATT to read
 * the payload (~200-500ms). Subsequent sightings of the same device skip
 * the GATT read and re-emit with updated RSSI from the cached payload.
 *
 * A staleness handler fires every 3 seconds. Devices not seen for 10 seconds
 * emit onDeviceLost and are purged from the cache.
 */
class BLEScanner(private val context: Context) {
  companion object {
    private const val TAG = "BLEScanner"
    private const val STALENESS_INTERVAL_MS = 3000L
    private const val STALENESS_THRESHOLD_MS = 10000L
  }

  var onDeviceFound: ((Map<String, Any?>) -> Unit)? = null
  var onDeviceLost: ((String) -> Unit)? = null

  private var leScanner: BluetoothLeScanner? = null
  private val handler = Handler(Looper.getMainLooper())

  // Devices that have been fully read (fast or slow path)
  private val readDevices = mutableMapOf<String, MutableMap<String, Any?>>()

  // Last seen timestamps for staleness tracking
  private val lastSeenTimestamps = mutableMapOf<String, Long>()

  // GATT connections in progress (held to prevent GC)
  private val pendingGattConnections = mutableMapOf<String, BluetoothGatt>()

  private val stalenessRunnable = object : Runnable {
    override fun run() {
      pruneStaleDevices()
      handler.postDelayed(this, STALENESS_INTERVAL_MS)
    }
  }

  private val scanCallback = object : ScanCallback() {
    override fun onScanResult(callbackType: Int, result: ScanResult?) {
      result?.let { handleScanResult(it) }
    }

    override fun onScanFailed(errorCode: Int) {
      Log.e(TAG, "Scan failed: errorCode=$errorCode")
    }
  }

  /** Starts scanning for Divot BLE advertisers. */
  @Suppress("MissingPermission")
  fun startScanning() {
    Log.i(TAG, "startScanning called")
    val bluetoothManager = context.getSystemService(Context.BLUETOOTH_SERVICE) as? BluetoothManager
    val adapter = bluetoothManager?.adapter
    if (adapter == null || !adapter.isEnabled) {
      Log.w(TAG, "Bluetooth adapter not available or not enabled")
      return
    }

    val scanner = adapter.bluetoothLeScanner
    if (scanner == null) {
      Log.e(TAG, "BLE scanner not available")
      return
    }
    leScanner = scanner

    val filter = ScanFilter.Builder()
      .setServiceUuid(BLEConstants.SERVICE_PARCEL_UUID)
      .build()

    val settings = ScanSettings.Builder()
      .setScanMode(ScanSettings.SCAN_MODE_LOW_LATENCY)
      .setReportDelay(0)
      .build()

    Log.i(TAG, "Starting BLE scan with service filter ${BLEConstants.SERVICE_UUID}")
    scanner.startScan(listOf(filter), settings, scanCallback)
    handler.postDelayed(stalenessRunnable, STALENESS_INTERVAL_MS)
  }

  /** Stops scanning, disconnects pending GATT connections, clears caches. */
  @Suppress("MissingPermission")
  fun stopScanning() {
    handler.removeCallbacks(stalenessRunnable)

    try {
      leScanner?.stopScan(scanCallback)
    } catch (e: Exception) {
      Log.w(TAG, "Error stopping scan", e)
    }
    leScanner = null

    for ((_, gatt) in pendingGattConnections) {
      try {
        gatt.disconnect()
        gatt.close()
      } catch (e: Exception) {
        Log.w(TAG, "Error closing GATT connection", e)
      }
    }

    pendingGattConnections.clear()
    readDevices.clear()
    lastSeenTimestamps.clear()
  }

  @Suppress("MissingPermission")
  private fun handleScanResult(result: ScanResult) {
    val device = result.device
    val deviceAddress = device.address
    val now = System.currentTimeMillis()
    val rssi = result.rssi

    // Capture device name now (may not be available inside GATT callback)
    val deviceName: String? = device.name

    lastSeenTimestamps[deviceAddress] = now

    // If already read, re-emit with updated RSSI (no reconnect)
    val cached = readDevices[deviceAddress]
    if (cached != null) {
      cached["rssi"] = rssi
      cached["lastSeen"] = now
      onDeviceFound?.invoke(cached)
      return
    }

    // Fast path: read service data from advertisement (Android advertiser)
    val serviceData = result.scanRecord?.getServiceData(BLEConstants.SERVICE_PARCEL_UUID)
    if (serviceData != null) {
      Log.d(TAG, "Fast path: service data found for $deviceAddress (${serviceData.size} bytes)")
      val parsed = BLEConstants.unpackPayload(serviceData)
      if (parsed != null) {
        val (platform, roomCode, _) = parsed
        Log.i(TAG, "Device found (fast): address=$deviceAddress, platform=$platform, roomCode=$roomCode, rssi=$rssi")
        val deviceInfo = mutableMapOf<String, Any?>(
          "id" to deviceAddress,
          "name" to deviceName,
          "platform" to platform,
          "roomCode" to roomCode,
          "rssi" to rssi,
          "lastSeen" to now,
        )
        readDevices[deviceAddress] = deviceInfo
        onDeviceFound?.invoke(deviceInfo)
        return
      }
    }

    // Slow path: GATT connect + characteristic read (iOS advertiser)
    if (pendingGattConnections.containsKey(deviceAddress)) return
    Log.i(TAG, "Slow path: connecting to $deviceAddress for GATT read (name=$deviceName, rssi=$rssi)")

    val gattCallback = object : BluetoothGattCallback() {
      override fun onConnectionStateChange(gatt: BluetoothGatt?, status: Int, newState: Int) {
        Log.d(TAG, "GATT onConnectionStateChange: address=$deviceAddress, status=$status, newState=$newState")
        if (newState == BluetoothProfile.STATE_CONNECTED) {
          gatt?.discoverServices()
        } else if (newState == BluetoothProfile.STATE_DISCONNECTED) {
          gatt?.close()
          pendingGattConnections.remove(deviceAddress)
        }
      }

      override fun onServicesDiscovered(gatt: BluetoothGatt?, status: Int) {
        val serviceUUIDs = gatt?.services?.map { it.uuid.toString() } ?: emptyList()
        Log.d(TAG, "GATT onServicesDiscovered: address=$deviceAddress, status=$status, services=$serviceUUIDs")
        if (status != BluetoothGatt.GATT_SUCCESS) {
          Log.w(TAG, "Service discovery failed (status=$status)")
          gatt?.disconnect()
          return
        }
        val service = gatt?.getService(BLEConstants.SERVICE_UUID)
        val characteristic = service?.getCharacteristic(BLEConstants.PAYLOAD_CHARACTERISTIC_UUID)
        if (characteristic != null) {
          Log.d(TAG, "Found our characteristic, reading value...")
          gatt.readCharacteristic(characteristic)
        } else {
          Log.w(TAG, "Service or characteristic not found (service=${service != null})")
          gatt?.disconnect()
        }
      }

      override fun onCharacteristicRead(
        gatt: BluetoothGatt?,
        characteristic: BluetoothGattCharacteristic?,
        status: Int
      ) {
        gatt?.disconnect()

        if (status != BluetoothGatt.GATT_SUCCESS) {
          Log.w(TAG, "Characteristic read failed (status=$status)")
          return
        }

        @Suppress("DEPRECATION")
        val value = characteristic?.value ?: return
        val parsed = BLEConstants.unpackPayload(value)
        if (parsed == null) {
          Log.w(TAG, "Failed to unpack payload (${value.size} bytes)")
          return
        }

        val (platform, roomCode, _) = parsed
        val currentRssi = rssi
        Log.i(TAG, "Device found (slow): address=$deviceAddress, platform=$platform, roomCode=$roomCode, rssi=$currentRssi")
        val deviceInfo = mutableMapOf<String, Any?>(
          "id" to deviceAddress,
          "name" to deviceName,
          "platform" to platform,
          "roomCode" to roomCode,
          "rssi" to currentRssi,
          "lastSeen" to System.currentTimeMillis(),
        )
        readDevices[deviceAddress] = deviceInfo
        lastSeenTimestamps[deviceAddress] = System.currentTimeMillis()
        onDeviceFound?.invoke(deviceInfo)
      }
    }

    val gatt = device.connectGatt(context, false, gattCallback)
    if (gatt != null) {
      pendingGattConnections[deviceAddress] = gatt
    }
  }

  @Suppress("MissingPermission")
  private fun pruneStaleDevices() {
    val now = System.currentTimeMillis()
    val staleIds = lastSeenTimestamps.filter { now - it.value > STALENESS_THRESHOLD_MS }.keys.toList()

    for (id in staleIds) {
      readDevices.remove(id)
      lastSeenTimestamps.remove(id)
      pendingGattConnections.remove(id)?.let { gatt ->
        try {
          gatt.disconnect()
          gatt.close()
        } catch (e: Exception) {
          Log.w(TAG, "Error closing stale GATT connection", e)
        }
      }
      onDeviceLost?.invoke(id)
    }
  }
}
